### ISSUE TITLE
API sketch

### ISSUE THREAD

**Issue by scottcorgan** [cmd-js/cmd#2]
Initial thoughts on API:

The basic idea is that everything is composable. You can compose flags onto commands, commands onto CLIs, flags on CLIs, executable functions onto anything, etc.

Ideally, even to the point of composing cli apps. This means that if I have a bunch of individual, tiny CLI apps, I can compose them together to make one larger app. Imagine being able to compose the following CLI apps to make one awesome, bigger cli app:

* https://github.com/sindresorhus/detect-indent-cli
* https://github.com/sindresorhus/find-up-cli
* etc ...

```js
var someCommand = command(
  name('dothis'),
  use(function (data, flags, options) {
    
  })
)

var someFlag = flag(
  name('-f', '--flag'),
  use(function (value, options) {
    
  })
);

var run = cli(
  use(function () {
    
    // this always runs before
  }),
  someCommand({/* pass options here */})
  someFlag({/* pass options here */}),
  use(function () {
    
    // This always runs after
  });
);

run(process.argv.slice(2));
```

**Comment by vsimonian** [comment]
Does this allow for commands to dynamically register plugins and endpoints?

e.g. if you have an app that wants to read metadata off a REST API to generate some commands, or if you're loading plugins arbitrarily out of a folder on a filesystem

```javascript
var app = cli(use(dynamicPluginLoader), ...)
```

What would `dynamicPluginLoader` have to look like in order to be able to append new plugins for use downstream? Is this possible without an `app.use(...)`?

**Comment by scottcorgan** [comment]
Note: composable plugins with options: https://gist.github.com/scottcorgan/3e7bd2bc6328e3c71e0a

**Comment by indexzero** [comment]
What is `use` in this scenario?

**Comment by indexzero** [comment]
I think writing writing your `help` plugin will be very telling how the composable model for `cmd` holds up since `help` in general has to be aware of all state to properly display the correct (and thus most helpful) error message.

A classic example of this is merging global and command flags.

**Comment by scottcorgan** [comment]
That's true. Where I've come it is a type of AST that gets parsed using the argv.

Thoughts on the composing API:

```js
var oneCommand = command(
  alias('me'),
  description('This is the "me" command and it does this and that'),
  example('command me <options>'),
  handler(function () {

    console.log('me command')
  }),
  handler(function () {

    console.log('another me command function')
  }),
  description('This is one command')
)

var flags = cli(f())

var program = cli(
  usage(
    description('cmd <command> <options>'),
    alias('--usage'),
    alias('-h', '--help')
  ),
  oneCommand({option: 'one'}),
  flags()
)

run(
	process.argv.slice(2),
	program()
)
```
